tianshou/utils/optim.py

```diff
@@ -20,3 +23,42 @@ def optim_step(
     if max_grad_norm:
         nn.utils.clip_grad_norm_(module.parameters(), max_norm=max_grad_norm)
     optim.step()
+
+
+_STANDARD_TORCH_OPTIMIZERS = [
+    torch.optim.Adam,
+    torch.optim.SGD,
+    torch.optim.RMSprop,
+    torch.optim.Adadelta,
+    torch.optim.AdamW,
+    torch.optim.Adamax,
+    torch.optim.NAdam,
+    torch.optim.SparseAdam,
+    torch.optim.LBFGS,
+]
+
+TOptim = TypeVar("TOptim", bound=torch.optim.Optimizer)
+
+
+def clone_optimizer(
```

Comment:
why do we need that?

Reply:
Because I relaxed the interface to pass `critic2=None, critic2_optim=None`. The default behavior is then to copy `critic` and `critic_optim`. The latter cannot be copied, hence this function.

Generally, I think we should bundle the nets and optimizers in a single abstraction, or let users pass optimizer factories instead. Currently, we always rely on the user to have a correct optimizer, i.e. we hope that `critic_optim` was indeed created with `critic.parameters()`. This is not verified at runtime, and if it doesn't hold, the algorithms won't work, of course.

Reply:
Indeed, I don't think this is needed (or elegant). The algorithm requires both critics and if the user wants to use the same type of critic twice, he can create two identical instances. This is more appropriate for a low-level interface than a perhaps contrived cloning mechanism.
In the high-level interface, which is currently being developed (#938), the user will be offered more convenient interfaces anyway (involving factories). For the more low-level interface, I would prefer to keep it simple and rather remove this.

Reply:
I disagree, in virtually all cases in tianshou the `net_2`, `critic2`, and `critic_optim2` are covered by the default. This will also be true for the absolute majority of users, I don't see the point in forcing them to create a second net/critic/optim if there is such a sensible default behavior. Of course, they are still able to do so if desired.

In my opinion, the main problem manifested here is that the nets and their optimizers are passed separately, as I outlined above, requiring this hacky and non-elegant solution. I view it as temporary step, until the "bundling" of optimizers and nets is improved - either by new abstractions of by switching to passing optim factories (like already done in many other torch-based libraries)

Reply:
I am also not aware of any recommendation from research about using different model architectures or learning rates for the dual-critic algorithms. There might be some exotic cases, of course

Reply:
IMO, clarity and simplicity should be prioritized over ease of use in a low-level interface. The high-level interface will solve the usability issues, so there is not really a good reason to solve them twice (in different ways).
It is just clearer to require exactly what the algorithm actually requires in the interface than to have a default which needs to be explained.

Reply:
I would be more inclined to agree with your notion if there was no high-level interface in development, mind you.

Reply:
There is value in maintaining a high degree of simplicity in the core (low-level) part of the library ([KISS](https://www.baeldung.com/cs/kiss-software-design-principle)).

Reply:
I don't think this one change leads to a slippery slope ;). The policy interface in a way is a high-level interface, and I suspect that researchers will continue using it as main entry point for a while.

However, I do see how a bundling of nets and optimizers (or use of factories) in the future might increase complexity if not done with much care, so I'm not 100% sure either.

@Trinkle23897 your call to give the decisive vote.

Reply:
I'm more inclined toward @opcode81. I agree with @MischaPanch that we should make a better interface, and I think 
> I view it as temporary step, until the "bundling" of optimizers and nets is improved - either by new abstractions of by switching to passing optim factories (like already done in many other torch-based libraries)

is the right direction we should push forward. As an intermediate step, if we don't care about clean/elegant, then the current stuff is totally fine; however, if we do care about it in this intermediate step, I tend to revert the optimizer change in this file, for two reasons:
- we will refactor this whole interface anyway in the future
- it's confusing people if the algorithm uses two critics/optimizers but the entrypoint only passes one critic and optimizer (that's why at the beginning I made the entrypoint to explicitly pass two critics instead of one)

Reply:
Since atm it's clearly marked as intermediate step and not to be used directly, I think it's fine to keep it for the time being. We should continue the discussion in a separate setting, probably only after the first version of the high-level interfaces has been merged. @opcode81 is preparing a Wiki entry for it. I will create an issue linking to this thread, so we keep it in mind.